### DISCUSSION TITLE
`relationJoins` default to `query`

### DISCUSSION THREAD

**Discussion by gapuchi** [prisma/prisma#25789] [Q&A] [answered]
### Question

Hi, I'm looking to use `relationJoins` in a production environment, but with the current potential problems of poor performance for certain queries, I wanted to add the plugin and keep the existing join behavior for all my queries.

That would enable be to opt in specific queries I wish to test without having to worry about the entire system. Is there a way for me to do this? AFAIK, adding the plugin switches all includes to use `join`

### How to reproduce (optional)

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by ujamX** [comment]
Hi, @gapuchi 

When the relationJoins plugin is enabled, all includes will automatically work in the join method. As far as I know, there is no setting to selectively apply the plugin behavior.

Why not create two PrismaClient instances, one with the RelationJoins plugin enabled and one with the default PrismaClient instance?
In this case, we expect that the RelationJoins plugin can only be applied to certain queries and the rest of the queries can retain their existing behavior.

**Reply by gapuchi** [reply]
Thanks @ujamX , creating two clients seems like a good idea. I'll try this approach. IIUC, we need to configure the plugin in the `schema.prisma` file. Does this mean I'll need two files, one with the plugin and one without? And have two different commands to generate the client?

**Reply by ujamX** [reply]
@gapuchi 
It seems to have been resolved while I was sleeping.
I'm glad you were resolve the issue.

**Comment by ludralph** [comment]
Hi @gapuchi 

You can control the join strategy on a per-query basis, which aligns with your goal of selectively testing joins without affecting your entire system.
When you enable the `relationJoins` preview feature, you can use the `relationLoadStrategy` option in your queries to specify whether you want to use a join or the original query-based approach. This allows you to opt-in to testing joins for specific queries while maintaining the existing behavior for others.

First, enable the relationJoins preview feature in your schema.prisma file:
```prisma
generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["relationJoins"]
}
```
Then, in your queries, you can use the `relationLoadStrategy` option to control the join behavior:
```js
const result = await prisma.user.findMany({
  relationLoadStrategy: 'query', // Use the original query-based approach
  include: {
    posts: true
  }
})
```
By setting `relationLoadStrategy: 'query'`, you're explicitly telling Prisma to use the original query-based approach for this specific query, even though the relationJoins feature is enabled.
For queries where you want to test the new join behavior, you can either omit the `relationLoadStrategy` option (as join is the default when the feature is enabled) or explicitly set it to 'join':
```js
const result = await prisma.user.findMany({
  relationLoadStrategy: 'join', // Use the new join-based approach
  include: {
    posts: true
  }
})
```

https://www.prisma.io/blog/prisma-orm-now-lets-you-choose-the-best-join-strategy-preview

**Reply by gapuchi** [reply]
I think the main concern here is that I'll need to go through the entire codebase to add `relationLoadStrategy: 'query', ` to my existing queries, which I'd like to avoid... I'll try ujamX's approach to create two clients.

On a higher level, once this is GA, is there any concern of changing the underlying behavior system wide? It seems to be safer if it was an opt-in change as opposed to opt-out

**Reply by ludralph** [reply, marked as answer]
I can see the benefit of @ujamX approach. It allows you to have more granular control over which queries use the new join strategy and which ones retain the existing behavior. For the two separate clients, you could do:
```js
import { PrismaClient } from '@prisma/client'

const prismaDefault = new PrismaClient()
const prismaWithJoins = new PrismaClient().$extends({
  query: {
    $allModels: {
      async $allOperations({ query, args }) {
        return query({ ...args, relationLoadStrategy: 'join' })
      },
    },
  },
})
```
You can then use `prismaDefault` for queries where you want to retain the existing behavior, and `prismaWithJoins` for queries where you want to test the new join strategy.

**Reply by gapuchi** [reply]
Thanks @ludralph! That clarifies how to create two different clients. Wasn't familiar with how to extend the client to do this.

**Reply by ludralph** [reply]
You're welcome. You can learn more about [Client Extensions](https://www.prisma.io/docs/orm/prisma-client/client-extensions).

**Reply by gapuchi** [reply]
Ended up with the following 

```typescript
const prisma = new PrismaClient().$extends({
  query: {
    $allModels: {
      async $allOperations({ query, args }) {
           return query({ relationLoadStrategy: 'query', ...args })
      },
    },
  },
})
```

By default, I'm setting the strategy to `query` to keep it consistent with before adding this feature and then providing the `args` after. This would allow me to override the `query` for specific queries.

**Reply by ludralph** [reply]
This looks good. I'm happy to see you got this resolved.

**Reply by gapuchi** [reply]
I had to update it slightly. We can't add the argument to all operations since that argument is not valid everywhere. I updated it to be inserted into specific operations.

```ts
const OPERATIONS_WITH_RELATION_JOIN = [
  'findUnique',
  'findUniqueOrThrow',
  'findFirst',
  'findFirstOrThrow',
  'findMany',
  'create',
  'delete',
  'update',
  'upsert',
]

...

 client.$extends({
      query: {
        $allModels: {
          async $allOperations({ query, args, operation }) {
            return OPERATIONS_WITH_RELATION_JOIN.includes(operation)
              ? query({ relationLoadStrategy: 'query', ...args })
              : query(args)
          },
        },
      },
    })
 ```
